### ISSUE TITLE
Too much memory usage of running tests

### ISSUE THREAD

**Issue by wgr523** [0xPolygonHermez/zkevm-prover#171]
Commit and tag: commit dc371981c6e0d7791aed2592b197539e04a188e4  tag: v0.4.5.0

**TL;DR**: I ran a proof generation test using `runFileGenProof` mode, the program used ~800GB virtual memory and ~250GB physical memory. I wonder could the memory usage be reduced?

Hi, I ran a proof generation test using `runFileGenProof` mode and input file `"inputFile": "input_executor.json",` I also removed the disabled config in `"cmPolsFile_disabled": "zkevm.commit",` to make it run. Then the program ran a while and output the following log:
```
../build/zkProver -c config_runFile.json
../build/zkProverVersion: v0.4.5.0
Number of cores=96
2022/11/07 14:52:08 --> WHOLE_PROCESS starting...
2022/11/07 14:52:08 --> LOAD_CONFIG_JSON starting...
Configuration:
    runFileGenProof=true
    proverServerPort=50051
    proverServerMockPort=50052
    proverClientPort=50051
    proverClientHost=127.0.0.1
    executorServerPort=50071
    executorClientPort=50071
    executorClientHost=127.0.0.1
    stateDBServerPort=50061
    stateDBURL=local
    inputFile=input_executor.json
    outputPath=output
    romFile=rom.json
    cmPolsFile=zkevm.commit
    cmPolsFileC12a=zkevm.c12a.commit
    cmPolsFileC12b=zkevm.c12b.commit
    constPolsFile=zkevm.const
    constPolsC12aFile=zkevm.c12a.const
    constantsTreeFile=zkevm.consttree
    constantsTreeC12aFile=zkevm.c12a.consttree
    starkFile=zkevm.prove.json
    starkFilec12a=zkevm.c12a.prove.json
    starkFilec12b=zkevm.c12b.prove.json
    starkZkIn=zkevm.proof.zkin.json
    starkZkInC12a=zkevm.c12a.zkin.proof.json
    starkZkInC12b=zkevm.c12b.zkin.proof.json
    verifierFile=zkevm.verifier.dat
    verifierFileC12a=zkevm.c12a.verifier.dat
    verifierFileC12b=zkevm.c12b.verifier.dat
    witnessFile=zkevm.witness.wtns
    witnessFileC12a=zkevm.c12a.witness.wtns
    witnessFileC12b=zkevm.c12a.witness.wtns
    starkVerifierFile=zkevm.g16.0001.zkey
    publicStarkFile=zkevm.public.json
    publicFile=public.json
    proofFile=proof.json
    keccakScriptFile=keccak_script.json
    keccakPolsFile=
    keccakConnectionsFile=keccak_connections.json
    storageRomFile=storage_sm_rom.json
    starkInfoFile=zkevm.starkinfo.json
    starkInfoC12aFile=zkevm.c12a.starkinfo.json
    databaseURL=local
    dbNodesTableName=
    dbProgramTableName=
    dbAsyncWrite=0
    cleanerPollingPeriod=600
    requestsPersistence=3600
    maxExecutorThreads=16
    maxProverThreads=16
    maxStateDBThreads=16
2022/11/07 14:52:08 <-- LOAD_CONFIG_JSON done: 0.000675 s
Output directory already exists
2022/11/07 14:52:08 --> PROVER_CONSTRUCTOR starting...
2022/11/07 14:52:08 --> ROM_LOAD starting...
ROM size: 4609 lines
2022/11/07 14:52:08 <-- ROM_LOAD done: 0.106546 s
2022/11/07 14:52:08 --> BINARY_EXECUTOR_BUILD_FACTORS starting...
2022/11/07 14:52:09 <-- BINARY_EXECUTOR_BUILD_FACTORS done: 0.582501 s
2022/11/07 14:52:09 --> KECCAK_F_SM_EXECUTOR_LOAD starting...
2022/11/07 14:52:10 <-- KECCAK_F_SM_EXECUTOR_LOAD done: 1.28128 s
2022/11/07 14:52:11 --> STARK_INFO_LOAD starting...
2022/11/07 14:52:12 <-- STARK_INFO_LOAD done: 1.79663 s
2022/11/07 14:52:13 --> LOAD_CONST_POLS_TO_MEMORY starting...
Stark::Stark() successfully copied 18387828736 bytes from constant file zkevm.const
2022/11/07 14:52:25 <-- LOAD_CONST_POLS_TO_MEMORY done: 11.944 s
2022/11/07 14:52:25 --> LOAD_CONST_TREE_TO_MEMORY starting...
Stark::Stark() successfully copied 37849399280 bytes from constant file zkevm.consttree
2022/11/07 14:52:50 <-- LOAD_CONST_TREE_TO_MEMORY done: 24.6647 s
2022/11/07 14:52:50 --> LOAD_CONST_POLS_2NS_TO_MEMORY starting...
2022/11/07 14:52:52 <-- LOAD_CONST_POLS_2NS_TO_MEMORY done: 2.89378 s
2022/11/07 14:52:52 --> COMPUTE_X_N_AND_X_2_NS starting...
2022/11/07 14:52:53 <-- COMPUTE_X_N_AND_X_2_NS done: 0.09648 s
2022/11/07 14:52:53 --> STARK_INFO_LOAD starting...
2022/11/07 14:52:53 <-- STARK_INFO_LOAD done: 0.080442 s
2022/11/07 14:52:53 --> LOAD_C12_A_CONST_POLS_TO_MEMORY starting...
StarkC12a::StarkC12a() successfully copied 2684354560 bytes from constant file zkevm.c12a.const
2022/11/07 14:52:55 <-- LOAD_C12_A_CONST_POLS_TO_MEMORY done: 2.62789 s
2022/11/07 14:52:55 --> LOAD_C12_A_CONST_TREE_TO_MEMORY starting...
StarkC12a::StarkC12a() successfully copied 15032385520 bytes from constant file zkevm.c12a.consttree
2022/11/07 14:53:10 <-- LOAD_C12_A_CONST_TREE_TO_MEMORY done: 14.7459 s
2022/11/07 14:53:10 --> LOAD_C12_A_CONST_POLS_2NS_TO_MEMORY starting...
2022/11/07 14:53:20 <-- LOAD_C12_A_CONST_POLS_2NS_TO_MEMORY done: 9.81749 s
2022/11/07 14:53:20 --> COMPUTE_X_N_AND_X_2_NS starting...
2022/11/07 14:53:21 <-- COMPUTE_X_N_AND_X_2_NS done: 0.680154 s
2022/11/07 14:53:21 --> STARK_INFO_LOAD starting...
2022/11/07 14:53:21 <-- STARK_INFO_LOAD done: 0.054227 s
2022/11/07 14:53:21 --> LOAD_C12_B_CONST_POLS_TO_MEMORY starting...
StarkC12b::StarkC12b() successfully copied 671088640 bytes from constant file zkevm.c12b.const
2022/11/07 14:53:21 <-- LOAD_C12_B_CONST_POLS_TO_MEMORY done: 0.58265 s
2022/11/07 14:53:21 --> LOAD_C12_B_CONST_TREE_TO_MEMORY starting...
StarkC12b::StarkC12b() successfully copied 3257016880 bytes from constant file zkevm.c12b.consttree
2022/11/07 14:53:23 <-- LOAD_C12_B_CONST_TREE_TO_MEMORY done: 2.09955 s
2022/11/07 14:53:23 --> LOAD_C12_B_CONST_POLS_2NS_TO_MEMORY starting...
2022/11/07 14:53:26 <-- LOAD_C12_B_CONST_POLS_2NS_TO_MEMORY done: 2.08549 s
2022/11/07 14:53:26 --> COMPUTE_X_N_AND_X_2_NS starting...
2022/11/07 14:53:26 <-- COMPUTE_X_N_AND_X_2_NS done: 0.162208 s
proverThread() started
cleanerThread() started
2022/11/07 14:53:32 <-- PROVER_CONSTRUCTOR done: 83.5699 s
2022/11/07 14:53:32 --> INPUT_LOAD starting...
2022/11/07 14:53:32 <-- INPUT_LOAD done: 0.000714 s
2022/11/07 14:53:32 --> PROVER_GEN_PROOF starting...
MEMORY INFO
MemTotal:               257574 MB
MemFree:               29473.2 MB
MemAvailable:           117313 MB
Buffers:               303.297 MB
Cached:                88674.4 MB
SwapCached:                  0 MB
SwapTotal:                2048 MB
SwapFree:                 2048 MB

PROCESS INFO
Pid:                     34912
User time:              166.98 s
Kernel time:               158 s
Total time:             324.98 s
Num threads:                98
Virtual mem:            146489 MB

Prover::genProof() timestamp: 20221107_065332_368691
Prover::genProof() UUID: dfe1da44-4369-4e93-9ab6-41cb3155d36e
Prover::genProof() input file: output/20221107_065332_368691_dfe1da44-4369-4e93-9ab6-41cb3155d36e.input_prover.json
Prover::genProof() public file: output/20221107_065332_368691_dfe1da44-4369-4e93-9ab6-41cb3155d36e.public.json
Prover::genProof() proof file: output/20221107_065332_368691_dfe1da44-4369-4e93-9ab6-41cb3155d36e.proof.json
Prover::genProof() successfully mapped 461239222272 bytes to file zkevm.commit
2022/11/07 14:53:32 --> EXECUTOR_EXECUTE starting...
2022/11/07 14:53:32 --> MAIN_EXECUTOR_EXECUTE starting...
2022/11/07 14:53:32 --> MAIN_EXECUTOR_EXECUTE starting...
2022/11/07 14:55:26 <-- MAIN_EXECUTOR_EXECUTE done: 114.118 s
MainExecutor::execute() done lastStep=8271 (0.098598%)
2022/11/07 14:55:26 <-- MAIN_EXECUTOR_EXECUTE done: 114.119 s
2022/11/07 14:55:26 --> PADDING_PG_SM_EXECUTE starting...
PaddingPGExecutor successfully processed 0 Poseidon hashes p=8388608 pDone=0 (0%)
2022/11/07 15:04:32 <-- PADDING_PG_SM_EXECUTE done: 546.403 s
2022/11/07 15:04:32 --> STORAGE_SM_EXECUTE starting...
StorageExecutor successfully processed 16 SMT actions (0.00944138%)
2022/11/07 15:14:46 <-- STORAGE_SM_EXECUTE done: 613.937 s
2022/11/07 15:14:46 --> BYTE4_SM_EXECUTE starting...
Byte4Executor successfully processed 265 Byte4 actions (0.00631809%)
2022/11/07 15:14:46 <-- BYTE4_SM_EXECUTE done: 0.02399 s
2022/11/07 15:14:46 --> ARITH_SM_EXECUTE starting...
ArithExecutor successfully processed 458 arith actions (0.174713%)
2022/11/07 15:14:47 <-- ARITH_SM_EXECUTE done: 1.13903 s
2022/11/07 15:14:47 --> BINARY_SM_EXECUTE starting...
BinaryExecutor successfully processed 743 binary actions (0.283432%)
2022/11/07 15:23:54 <-- BINARY_SM_EXECUTE done: 546.096 s
2022/11/07 15:23:54 --> MEM_ALIGN_SM_EXECUTE starting...
MemAlignExecutor successfully processed 0 memory align actions (0%)
2022/11/07 15:33:21 <-- MEM_ALIGN_SM_EXECUTE done: 567.836 s
2022/11/07 15:33:21 --> MEMORY_SM_EXECUTE starting...
2022/11/07 15:33:21 --> MEMORY_EXECUTOR_REORDER starting...
2022/11/07 15:33:21 <-- MEMORY_EXECUTOR_REORDER done: 0.009024 s
MemoryExecutor successfully processed 5249 memory accesses (0.062573%)
2022/11/07 15:43:24 <-- MEMORY_SM_EXECUTE done: 602.24 s
2022/11/07 15:43:24 --> PADDING_KK_SM_EXECUTE starting...
PaddingKKExecutor successfully processed 6 Keccak hashes p=8388608 pDone=1088 (0.01297%)
2022/11/07 15:54:10 <-- PADDING_KK_SM_EXECUTE done: 645.868 s
2022/11/07 15:54:10 --> PADDING_KK_BIT_SM_EXECUTE starting...
PaddingKKBitExecutor successfully processed 468 Keccak actions p=932724 pDone=932724 (11.1189%)
2022/11/07 15:55:14 <-- PADDING_KK_BIT_SM_EXECUTE done: 64.2091 s
2022/11/07 15:55:14 --> NINE2ONE_SM_EXECUTE starting...
Nine2OneExecutor successfully processed 468 Keccak hashes (98.2015%)
2022/11/07 15:56:06 <-- NINE2ONE_SM_EXECUTE done: 52.2503 s
2022/11/07 15:56:06 --> KECCAK_F_SM_EXECUTE starting...
KeccakFExecutor successfully processed 52 Keccak-F actions (98.2015%)
2022/11/07 16:03:33 <-- KECCAK_F_SM_EXECUTE done: 446.936 s
2022/11/07 16:03:33 --> NORM_GATE_9_SM_EXECUTE starting...
NormGate9Executor successfully processed 2745912 NormGate9 actions (98.2015%)
2022/11/07 16:13:12 <-- NORM_GATE_9_SM_EXECUTE done: 578.956 s
2022/11/07 16:13:12 --> POSEIDON_G_SM_EXECUTE starting...
PoseidonGExecutor successfully processed 139 Poseidon hashes p=8388608 pDone=4309 (0.0513673%)
2022/11/07 16:22:58 <-- POSEIDON_G_SM_EXECUTE done: 586.196 s
2022/11/07 16:22:58 <-- EXECUTOR_EXECUTE done: 5366.22 s
2022/11/07 16:22:58 --> SAVE_PUBLICS_JSON starting...
2022/11/07 16:22:58 <-- SAVE_PUBLICS_JSON done: 0.01256 s
2022/11/07 16:22:58 --> STARK_PROOF starting...
2022/11/07 16:22:59 --> STARK_STEP_1 starting...
2022/11/07 16:22:59 --> STARK_STEP_1_LDE_AND_MERKLETREE starting...
2022/11/07 16:22:59 --> STARK_STEP_1_LDE starting...
[1]    34912 killed     ../build/zkProver -c config_runFile.json
```
It turned out that the program used too much memory and got killed. Here is the result of command `sudo dmesg`:
```
[343569.548118] Out of memory: Killed process 34912 (zkProver) total-vm:852092648kB, anon-rss:259203836kB, file-rss:0kB, shmem-rss:0kB, 
```
This means the program used ~800GB virtual memory and ~250GB physical memory. I wonder could the memory usage be reduced?

**Comment by lightscale-luke** [comment]
May I ask you how did you get to run the test?
I took following steps listed on README.md

1. Clone repository
2. Compile

to run the test, I ran 
`$ ../build/zkProver -c config_runFile.json `

Since I couldn't find the config_runFile.json under testvectors ditrectory,
I copied from config/config_prover.json and add 'runFileGenProof : true" to the file instead.

then when I ran 
`../build/zkProver -c config_runFile.json `

I get such messages like
    ```
[directory-path]/recursive1/recursive1.const does not exist.
    [directory-path]/recursive2/recursive2.const does not exist.
```
    and so on. 


Am i missing something or should I set configuration differently?
